benchmark/rocblas/blas1/asum.cpp

```diff
@@ -0,0 +1,145 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename asum.cpp
+ *
+ **************************************************************************/
+
+#include "../utils.hpp"
+
+template <typename scalar_t>
+std::string get_name(int size) {
+  std::ostringstream str{};
+  str << "BM_Asum<" << blas_benchmark::utils::get_type_name<scalar_t>() << ">/";
+  str << size;
+  return str.str();
+}
+
+template <typename scalar_t, typename... args_t>
+static inline void rocblas_asum_f(args_t&&... args) {
+  if constexpr (std::is_same_v<scalar_t, float>) {
+    CHECK_ROCBLAS_STATUS(rocblas_sasum(std::forward<args_t>(args)...));
+  } else if constexpr (std::is_same_v<scalar_t, double>) {
+    CHECK_ROCBLAS_STATUS(rocblas_dasum(std::forward<args_t>(args)...));
+  }
+  return;
+}
+
+template <typename scalar_t>
+void run(benchmark::State& state, rocblas_handle& rb_handle, index_t size,
+         bool* success) {
+  // Google-benchmark counters are double.
+  double size_d = static_cast<double>(size);
+  state.counters["size"] = size_d;
+  state.counters["n_fl_ops"] = 2 * size_d;
+  state.counters["bytes_processed"] = size_d * sizeof(scalar_t);
+
+  using data_t = scalar_t;
+
+  // Create data
+  std::vector<data_t> v1 = blas_benchmark::utils::random_data<data_t>(size);
+  scalar_t vr;
+
+  {
+    // Device memory allocation & H2D Copy
+    blas_benchmark::utils::HIPVector<scalar_t> d_v1(size, v1.data());
+    blas_benchmark::utils::HIPScalar<scalar_t> d_vr(vr);
+
+#ifdef BLAS_VERIFY_BENCHMARK
+    // Reference asum
+    data_t vr_ref = reference_blas::asum(size, v1.data(), 1);
+
+    // Rocblas verification asum
+    data_t vr_temp = 0;
+
+    {
+      // Temp result on device (copied back to Host upon destruction)
+      blas_benchmark::utils::HIPScalar<scalar_t, true> d_vr_temp(vr_temp);
+      // rocBLAS function call
+      rocblas_asum_f<scalar_t>(rb_handle, size, d_v1, 1, d_vr_temp);
+    }
+
+    if (!utils::almost_equal(vr_temp, vr_ref)) {
+      std::ostringstream err_stream;
+      err_stream << "Value mismatch: " << vr_temp << "; expected " << vr_ref;
+      const std::string& err_str = err_stream.str();
+      state.SkipWithError(err_str.c_str());
+      *success = false;
+    };
+#endif
+
+    auto blas_warmup = [&]() -> void {
+      rocblas_asum_f<scalar_t>(rb_handle, size, d_v1, 1, d_vr);
+      CHECK_HIP_ERROR(hipStreamSynchronize(NULL));
```

Comment:
in cublas we moved this `streamSynchronize` after the `warmup`. Could you do the same?

Reply:
Sure will do!